### PR TITLE
Added top n log probs

### PR DESCRIPTION
**Pull request by shanmugamr1992**
# What does this PR do ?

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 1]
```diff
@@ -722,6 +736,54 @@ def _dynamic_step_calculate_log_probs(self, logits: Tensor) -> Optional[Tensor]:
...
+
+        active_request_count = context.total_request_count - context.paused_request_count
+
+        # Get logits - handle both decode (last token only) and prefill (all tokens) modes
+        if materialize_only_last_token_logits:
+            # Decode mode: only last token logits
+            # Shape: [1, active_request_count, vocab_size] -> [active_request_count, vocab_size]
+            last_token_logits = logits.squeeze(0)
+            # Compute log probabilities
+            log_probs = torch.nn.functional.log_softmax(last_token_logits, dim=-1).to(torch.float32)
```
Comment (tdene) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: `log_softmax` is also calculated in `dynamic_context.py`, under `calculate_log_probs`. It would be good if that calculation can be reused.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 2]
```diff
@@ -722,6 +736,54 @@ def _dynamic_step_calculate_log_probs(self, logits: Tensor) -> Optional[Tensor]:
...
+                top_n = int(self.top_n_logprobs_cuda[req_idx].item())
+                if top_n > 0:
+                    # Get top-n logprobs and indices for this request (single token)
+                    top_n_logits = torch.topk(log_probs[req_idx], k=min(top_n, log_probs.size(-1)))
+                    top_n_results[req_idx] = [(top_n_logits.values, top_n_logits.indices)]
+        else:
+            # Prefill/non-decode mode: all tokens
+            # Shape: [1, num_tokens, vocab_size]
+            # Need to extract per-request logits using last_token_logits helper
+            last_token_logits = context.last_token_logits(logits)  # Shape: [active_request_count, vocab_size]
```
Comment (tdene) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: You can make the code shorter by starting with a
```
        # Last token logits.
        if materialize_only_last_token_logits:
            last_token_logits = logits.squeeze(0)
        else:
            last_token_logits = context.last_token_logits(logits)
```

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 3]
```diff
@@ -722,6 +736,54 @@ def _dynamic_step_calculate_log_probs(self, logits: Tensor) -> Optional[Tensor]:
...
+            
+            # Compute log probabilities for last tokens
```
Comment (tdene) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: This misses logprobs for prefill.

This is why I would reuse the implementation in `dynamic_context.py`, under `calculate_log_probs`. That accounts for prefill logprobs, and the logic to do so is non-trivial.

**Comment by kvareddy** [comment]
@santhnm2 can you please take a look at this MR?

File: megatron/core/inference/contexts/dynamic_context.py
Comment (santhnm2) [inline megatron/core/inference/contexts/dynamic_context.py:1759]: Can you update the docstring comment to indicate we're also returning the raw logprobs for potential top-n-logprobs calculations?

File: megatron/core/inference/engines/dynamic_engine.py [hunk 4]
```diff
@@ -630,7 +672,49 @@ def post_process_requests(
...
+                        # append to prompt_top_n_logprobs. Otherwise append to generated_top_n_logprobs.
+                        if (
+                            request.sampling_params.return_prompt_top_n_logprobs
+                            and total_accumulated < prompt_length - 1
+                        ):
+                            request.prompt_top_n_logprobs.append(logit_dict)
+                        else:
+                            request.generated_top_n_logprobs.append(logit_dict)
+
+                active_requests.append(request)
```
Comment (santhnm2) [inline megatron/core/inference/engines/dynamic_engine.py]: Can this be generalized with the code block above in the non-chunked prefill case?

Comment (shanmugamr1992) [reply megatron/core/inference/engines/dynamic_engine.py]: done

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 5]
```diff
@@ -391,9 +391,19 @@ def _add_request():
...
                         finished_request.prompt_log_probs + finished_request.generated_log_probs
                     )
+                if finished_request.sampling_params.top_n_logprobs > 0:
+                    request.generated_top_n_logprobs = getattr(
+                        finished_request, 'generated_top_n_logprobs', None
+                    )
+                if finished_request.sampling_params.return_prompt_top_n_logprobs:
+                    request.prompt_top_n_logprobs = getattr(
+                        finished_request, 'prompt_top_n_logprobs', None
+                    )
```
Comment (tdene) [inline examples/inference/gpt/gpt_dynamic_inference.py]: Personally, I strongly oppose conditional attributes that make us have to use `getattr`. I do not see a reason why we cannot just put `generated_top_n_logprobs` and `prompt_top_n_logprobs` directly into the `InferenceRequest` dataclass.

It is more confusing for the `InferenceRequest` object to gain unlisted attributes over time, than it is for us to list all the attributes ahead of time - even if not all flows will need them.

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 6]
```diff
@@ -436,9 +446,12 @@ def main():
         temperature=args.temperature,
         top_k=args.top_k,
         top_p=args.top_p,
+        skip_prompt_log_probs=args.skip_prompt_log_probs,
         return_log_probs=args.return_log_probs,
         num_tokens_to_generate=args.num_tokens_to_generate,
         termination_id=args.termination_id if args.termination_id is not None else tokenizer.eod,
+        top_n_logprobs=args.top_n_logprobs,
+        return_prompt_top_n_logprobs=args.return_prompt_top_n_logprobs,
```
Comment (tdene) [inline examples/inference/gpt/gpt_dynamic_inference.py:435]: nit, but I think that we should be consistent with our naming. The rest of the codebase uses `log_probs`. Newly added attributes should as well, instead of `logprobs`.

There's several examples of `logprobs` in this PR.

File: examples/inference/gpt/gpt_dynamic_inference.py [hunk 7]
```diff
@@ -564,13 +577,16 @@ def escape_str(s):
...
+                    print(f' Attributes of request {i}: {req.__dict__}')
                     result_dict = {
                         "input_prompt": req.prompt_text,
                         "generated_text": req.output_text,
                         "generated_tokens": req.output_tokens,
                         "latency": req.time_end - req.time_start,
                         "cuda_graph_request_count_map" : result["cuda_graph_request_count_map"],
                         "step_count" : engine.step_count,
+                        "top_n_logprobs" : getattr(req, 'generated_top_n_logprobs', None),
+                        "prompt_top_n_logprobs" : getattr(req, 'prompt_top_n_logprobs', None),
```
Comment (tdene) [inline examples/inference/gpt/gpt_dynamic_inference.py:561]: See my comment [here](https://github.com/NVIDIA/Megatron-LM/pull/2262/files#r2553273196) about `getattr`, versus having the attributes be explicitly listed in the dataclass.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 8]
```diff
@@ -699,28 +710,131 @@ def _dynamic_step_log_probs_bookkeeping(self) -> bool:
...
         active_request_count = context.total_request_count - context.paused_request_count
 
-        ret = context.calculate_log_probs(
+        ret, log_probs = context.calculate_log_probs(
             logits,
             self.sampled_tokens_cuda[:active_request_count],
             only_last_token_logits=materialize_only_last_token_logits,
         )
-        return ret
+        return ret, log_probs
```
Comment (tdene) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: nit: just `return context.calculate_log_probs` directly, without the need for `ret, log_probs` variables.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py [hunk 9]
```diff
@@ -814,10 +925,26 @@ async def async_generate_output_tokens_dynamic_batch(
...
 
         return_log_probs = self._dynamic_step_log_probs_bookkeeping()
-        if return_log_probs:
-            log_probs = self._dynamic_step_calculate_log_probs(logits)
+        return_top_n_logprobs = self._dynamic_step_top_n_logprobs_bookkeeping()
+
+        # Calculate log probs if needed for return or for top-n calculation.
+        # Per the constraint in dynamic_engine.py, any request with top_n_logprobs > 0
+        # must also have return_log_probs=True. However, during prefill with skip_prompt_log_probs,
+        # we might not store prompt log probs but still need them for top-n calculation.
```
Comment (tdene) [inline megatron/core/inference/text_generation_controllers/text_generation_controller.py]: I don't understand this comment. Are we saying that `skip_prompt_log_probs` should skip returning the normal prompt log probs, but it should not skip returning top n prompt log probs?

**Comment by tdene** [comment]
I left a few nit comments, but the general code looks good to me.

I will approve this after main is merged into it.

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 5 above)
Comment (shanmugamr1992) [reply examples/inference/gpt/gpt_dynamic_inference.py]: I agree. Will change it.

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 6 above)
Comment (shanmugamr1992) [reply examples/inference/gpt/gpt_dynamic_inference.py:435]: I agree. However the static engine and all the other parts have already used this logprobs. So it becomes very cumbersome to change this because backward compatiblity tests will break.

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 7 above)
Comment (shanmugamr1992) [reply examples/inference/gpt/gpt_dynamic_inference.py:561]: Done

File: examples/inference/gpt/gpt_dynamic_inference.py (hunk 6 above)
Comment (tdene) [reply examples/inference/gpt/gpt_dynamic_inference.py:435]: Sounds good.

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py (hunk 8 above)
Comment (shanmugamr1992) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py]: Done

File: megatron/core/inference/text_generation_controllers/text_generation_controller.py (hunk 9 above)
Comment (shanmugamr1992) [reply megatron/core/inference/text_generation_controllers/text_generation_controller.py]: Yeah i agree, it was confusing. I simplified the changes. so its as follows 
1. If you want top n logprobs you have to set return log probs to true. 
Then that will take you into the condition to calculate log probs. and this calculated log probs will be used to get the top n logprobs. 
2. In dynamic engine the calculated log probs will be used to either put in the prompt log probs or the generated log probs.

File: megatron/core/inference/contexts/dynamic_context.py [hunk 10]
```diff
@@ -1749,7 +1749,7 @@ def update_requests(self, active_requests_mask: Tensor, new_tokens: Tensor) -> T
 
     def calculate_log_probs(
         self, logits: Tensor, new_tokens: Tensor, only_last_token_logits: Optional[bool] = False
-    ) -> List[List[float]]:
+    ) -> Tuple[List[List[float]], Tensor]:
```
Comment (ericharper) [inline megatron/core/inference/contexts/dynamic_context.py:1748]: This is a breaking change, will this break any downstream code or is this private?

Comment (tdene) [reply megatron/core/inference/contexts/dynamic_context.py:1748]: This is a good point that I missed.

Subscribing to this thread; I am also very interested in the answer to this, since I intended to optimize the `calculate_log_probs` method in 2 PRs' time (circa Dec 10th?).

@wdykas I think you are the main author of this method. Do you know how much (if any) downstream dependence there should be on it?

Comment (shanmugamr1992) [reply megatron/core/inference/contexts/dynamic_context.py:1748]: This is only used by internal code. So shouldnt be a problem.

**Comment by shanmugamr1992** [comment]
/ok to test [300c130](https://github.com/NVIDIA/Megatron-LM/pull/2262/commits/300c1300741222dd47855f794a708e2c852c3b1e)

**Comment by shanmugamr1992** [comment]
/ok to test [4303dd0](https://github.com/NVIDIA/Megatron-LM/pull/2262/commits/4303dd007e996cddd8f5ddf882a4ca6b0c58d484)

**Comment by shanmugamr1992** [comment]
/ok to test [0740fd6](https://github.com/NVIDIA/Megatron-LM/pull/2262/commits/0740fd621c01fb99c0e1dc9e789a8405a7794e54)
